pkg/proxy/apis/config/validation/validation.go

```diff
@@ -69,14 +71,20 @@ func Validate(config *kubeproxyconfig.KubeProxyConfiguration) field.ErrorList {
 	allErrs = append(allErrs, validateHostPort(config.MetricsBindAddress, newPath.Child("MetricsBindAddress"))...)
 
 	if config.ClusterCIDR != "" {
-		if config.FeatureGates[string(kubefeatures.IPv6DualStack)] {
-			cidrs := strings.Split(config.ClusterCIDR, ",")
-			for _, cidr := range cidrs {
-				if _, _, err := net.ParseCIDR(cidr); err != nil {
-					allErrs = append(allErrs, field.Invalid(newPath.Child("ClusterCIDR"), cidr, "must be a valid CIDR block (e.g. 10.100.0.0/16 or FD02::0:0:0/96)"))
-				}
+		cidrs := strings.Split(config.ClusterCIDR, ",")
+
+		switch {
+		// if DualStack and more than one cidr validate if there is at least one of each IP family
+		case utilfeature.DefaultFeatureGate.Enabled(kubefeatures.IPv6DualStack) && len(cidrs) > 1:
```

Comment:
hmm... this is checking whether the feature is enabled, but the config we're validating also can **set** feature enablement, and I don't think the global state has been set from that at the point this is validated

we might need to do something like this at the top of this method:
```
effectiveFeatures := utilfeature.DefaultFeatureGate.DeepCopy()
if err := effectiveFeatures.SetFromMap(config.FeatureGates); err != nil {
  allErrs = append(allErrs, field.Invalid(newPath.Child("featureGates"), config.FeatureGates, err.Error()))
}
```

and then use effectiveFeatures here to check feature enablement

Reply:
I was struggling with this a long time, and my understanding from the code is that this validation is called after merging this configuration and the flags and setting the featureGates

https://github.com/kubernetes/kubernetes/blob/2808d7154dc820c2718d92eb2f000ffc5840b5d1/cmd/kube-proxy/app/server.go#L486-L491

opts.Complete() returns with

https://github.com/kubernetes/kubernetes/blob/2808d7154dc820c2718d92eb2f000ffc5840b5d1/cmd/kube-proxy/app/server.go#L244

however, I see that this is too brittle and your approach will prevent future mistakes